Fix keypath-as-function crasher

beccadax opened the pull request:
The autoclosures generated for the keypath-as-function feature were not added to the list of closures that needed captures computed. In top-level code, this caused a crash. Fixes rdar://problem/56055600.

beccadax committed:
Fix keypath-as-function crasher

The autoclosures generated for the keypath-as-function feature were not added to the list of closures that needed captures computed. In top-level code, this caused a crash. Fixes rdar://problem/56055600.

brentdax commented:
I'm trying to see if I can get an assertion in that would have caught this bug, but let's test the fix.

@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test linux platform

beccadax committed:
[SILGen] Assert that captures have been computed

Before using the capture info, SILGen now asserts that it has computed, except for declarations which are not in a local context and therefore can’t have captures.

This causes failures in REPL tests, but they are actual failures—there’s a bug in TypeCheckREPL.

slavapestov commented on lib/SIL/TypeLowering.cpp:
```diff
@@ -2109,8 +2111,17 @@ TypeConverter::hasLoweredLocalCaptures(SILDeclRef fn) {
   return !getLoweredLocalCaptures(fn).getCaptures().empty();
 }
 
+static bool cannotCapture(DeclContext *DC) {
```
Since this appears twice it should be a method on the TypeLowering object. Also it should take a SILDeclRef or AnyFunctionRef

brentdax replied:
Yes, the duplication in the checks is the main thing I need to clean up (if this is the correct check).

slavapestov commented on lib/SILGen/SILGenProlog.cpp:
```diff
@@ -445,6 +451,8 @@ void SILGenFunction::emitProlog(const CaptureInfo &captureInfo,
 
   // Emit the capture argument variables. These are placed last because they
   // become the first curry level of the SIL function.
+  assert((captureInfo.hasBeenComputed() || cannotCapture(DC)) &&
```
You can early return and skip the whole loop below if cannotCapture() is false.

brentdax replied:
I don't think it'll make a real difference because the loop will exit very quickly anyway, and to be honest I don't trust `canCaptureFromParent()` enough to risk a subtle miscompile from it.

slavapestov commented on validation-test/compiler_crashers_2_fixed/0209-rdar56055600.swift:
```diff
@@ -0,0 +1,2 @@
+// RUN: %target-swift-frontend %s -emit-silgen -o /dev/null
+let _: ([Int]) -> Int = \[Int].count
```
A FileCheck test in test/SILGen/ might be better since then you can CHECK: that the closure's SIL function has the capture as a parameter.

beccadax committed:
Compute captures of REPL closures

REPLChecker::generatePrintOfExpression() type-checked the closures it generated, but did not add them to the ClosuresWithUncomputedCaptures list.

brentdax commented:
@swift-ci please smoke test

beccadax committed:
[NFC] Base emitGeneratorFunction() on canCaptureFromParent()

beccadax committed:
Merge branch 'master' into everything-evil

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test
